### ISSUE TITLE
TGraph bin labels are set incorrectly

### ISSUE THREAD

**Issue by mikhailremnev** [root-project/jsroot#297]
## Bug description

Consider the TGraph object created using the following code:
```cpp
  auto g = new TGraph();
  g->SetTitle("y = -x");
  g->SetMarkerStyle(kFullCircle);

  for (unsigned short i = 0; i < 21; i++) g->SetPoint(i, i, -i);
  for (unsigned short i = 0; i < 21; i++) g->GetXaxis()->SetBinLabel(i + 1, std::to_string(i + 1).c_str());
  g->GetXaxis()->Set(21, 0, 21);
```
(the number 21 is arbitrary, the issue can be seen with any number of bins larger than 10)

Drawn with ROOT library, the plot looks like this:
![image](https://github.com/root-project/jsroot/assets/89205155/28043289-a993-4ef3-924e-46eba57ae8e1)

However, if I draw the same TGraph object in JSROOT (version >= 5.9.1), it looks like this:
![image](https://github.com/root-project/jsroot/assets/89205155/cc5e3919-6be9-4f96-b801-751e3e39b387)

It can be seen that the points are squeezed from (0, 20) range to (0, 19) range.

## Reason (to my superficial understanding)

If we comment out the line that sets labels:
```cpp
  // for (unsigned short i = 0; i < 21; i++) g->GetXaxis()->SetBinLabel(i + 1, std::to_string(i + 1).c_str());
```

We can see that in versions <= 5.9.0, the X range is set to (0, 20.5). And it is set to (0, 22) in higher versions (I've tested 6.0.0, 6.2.0, 6.3.4, 7.0.0, 7.6.0).
![image](https://github.com/root-project/jsroot/assets/89205155/3d82f3fd-7a54-45f0-b272-c44d4627f5d8)

So maybe binning of underlying histogram is somewhat off?

I have verified that the bug is caused by commit [6c1aa](https://github.com/root-project/jsroot/commit/6c1aaae03823f76c28e31176a734b346d1cd2bc6).

Specifically, it is caused by the removal of these lines:
```javascript
      else {
         // make logic like in the TGraph::GetHistogram
         if (!graph.TestBit(kResetHisto)) return histo;
         graph.InvertBit(kResetHisto);
      }
```

This patch fixes the histogram but **probably breaks other stuff**:
```diff
diff --git a/modules/hist2d/TGraphPainter.mjs b/modules/hist2d/TGraphPainter.mjs
index 1b4e9a42a..e4d7d819c 100644
--- a/modules/hist2d/TGraphPainter.mjs
+++ b/modules/hist2d/TGraphPainter.mjs
@@ -332,6 +332,11 @@ class TGraphPainter extends ObjectPainter {
          this._own_histogram = true;
          this.setHistogram(histo);
       } else if ((histo.fMaximum !== kNoZoom) && (histo.fMinimum !== kNoZoom)) {
+         var kResetHisto = BIT(17);   ///< fHistogram must be reset in GetHistogram
+         // make logic like in the TGraph::GetHistogram
+         if (!graph.TestBit(kResetHisto)) return histo;
+         graph.InvertBit(kResetHisto);
+
          minimum = histo.fMinimum;
          maximum = histo.fMaximum;
       }
```

**Comment by linev** [comment]
Hi,

Thanks for reporting the problem.
I can reproduce it and will try to find a fix for it.

Sorry for long delay - I was in long vacations.

**Comment by linev** [comment]
I applied different approach.

If there are custom labels - histogram axis will not be modified.
It has to be properly configured as in your example.

Fix is in the master, 7.5 and 7.6 branch.

Do you need it in the released version?

**Comment by mikhailremnev** [comment]
@linev thank you very much for fixing this bug.

I think we are fine with running the local build of master branch.
(in my opinion this issue can be closed, not sure if you'd like to keep it open or not, any way is fine with me)
